### PR TITLE
Fix clamp boundary subgradients

### LINKED ISSUE #184572: `torch.maximum` / `torch.clamp_min` gradients at tie point violate PyTorch's documented subgradient convention and are inconsistent with `F.relu`
### 🐛 Describe the bug

`torch.maximum(x, tensor(0))`, `torch.clamp_min(x, 0)`, and `torch.clamp(x, min=0)` all implement `max(x, 0)`, but their gradients at `x=0` contradict PyTorch's own autograd documentation and are mutually inconsistent.

## Minimal reproducer

```python
import torch
import torch.nn.functional as F

formulations = [
    ("F.relu(x)",                   lambda x: F.relu(x)),
    ("torch.maximum(x, tensor(0))", lambda x: torch.maximum(x, torch.tensor(0.0))),
    ("torch.clamp_min(x, 0)",       lambda x: torch.clamp_min(x, 0.0)),
    ("torch.clamp(x, min=0)",       lambda x: torch.clamp(x, min=0.0)),
]

for name, fn in formulations:
    x = torch.tensor(0.0, requires_grad=True)
    fn(x).backward()
    print(f"{name:<35} grad at x=0: {x.grad.item()}")
```

## Observed output

```
F.relu(x)                           grad at x=0: 0.0
torch.maximum(x, tensor(0))         grad at x=0: 0.5
torch.clamp_min(x, 0)               grad at x=0: 1.0
torch.clamp(x, min=0)               grad at x=0: 1.0
```

## Expected output

```
F.relu(x)                           grad at x=0: 0.0
torch.maximum(x, tensor(0))         grad at x=0: 0.0
torch.clamp_min(x, 0)               grad at x=0: 0.0
torch.clamp(x, min=0)               grad at x=0: 0.0
```

## Why 0.0 is the correct answer

All four expressions compute the same mathematical function `max(x, 0)`, yet they produce three different gradients at `x=0`. This is a problem in practice: users who refactor between these equivalent formulations silently get different gradient behaviour at zero-initialised weights or ReLU activations that land exactly on zero.

PyTorch's own [autograd documentation](https://pytorch.org/docs/stable/notes/autograd.html) specifies:

> *"If the function is convex (at least locally), use the sub-gradient of minimum norm."*

`max(x, 0)` is convex. Its subdifferential at `x=0` is the interval `[0, 1]`. The element
of minimum norm is **0.0**.

| Formulation | grad at x=0 | Consistent with docs? |
|---|---|---|
| `F.relu(x)` | 0.0 | ✓ |
| `torch.abs(x)` | 0.0 | ✓ |
| `torch.maximum(x, tensor(0))` | 0.5 | ✗ |
| `torch.clamp_min(x, 0)` | 1.0 | ✗ |
| `torch.clamp(x, min=0)` | 1.0 | ✗ |

`F.relu` and `torch.abs` already follow the documented minimum-norm convention. `torch.maximum` and `torch.clamp` do not.

### Versions

```
PyTorch version: 2.13.0.dev20260512+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-37-generic-x86_64-with-glibc2.39
Is CUDA available: True
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 5090
Nvidia driver version: 590.48.01

[pip3] numpy==2.4.4
[pip3] torch==2.13.0.dev20260512+cu130
[pip3] triton==3.7.0+git88b227e2
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by HU-qingqing**
### summary
Use minimum-norm subgradients for clamp boundaries. Scalar bounds now give zero  self-gradient at the boundary, while one-sided Tensor bounds split ties evenly between self and the bound. Leave two-Tensor-bound  clamp unchanged.

This PR updates clamp boundary gradients to follow the minimum-norm subgradient convention.

Earlier attempts only changed scalar-bound/one-sided Tensor bounds clamp cases to return zero gradient at the boundary. Based on the issue discussion, Tensor bounds need to be handled separately: one-sided Tensor bounds are multivariate, so ties should split the gradient evenly between `self` and the bound, matching the `maximum` / `minimum` convention.

This PR intentionally does not change the two-Tensor-bound case: Applying the minimum-norm rule there would require more boundary/tie handling and additional tensor ops in a performance-sensitive path, so this PR keeps that case unchanged. I would like to discuss the two-Tensor-bound case separately, especially the expected behavior for degenerate/tie cases such as `self == min == max`, and whether the extra boundary handling is worth the potential performance cost.

### Test Plan:
python test/test_autograd.py TestAutogradCPU.test_clamp_boundary_subgradient_cpu
python test/test_autograd.py TestAutogradCPU.test_clamp_boundary_jvp_cpu
python test/test_autograd.py TestAutogradCUDA.test_clamp_boundary_subgradient_cuda
python test/test_autograd.py TestAutogradCUDA.test_clamp_boundary_jvp_cuda

fixes #184572

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -430,22 +434,22 @@
...
 - name: clamp_min(Tensor self, Scalar min) -> Tensor
-  self: where(self >= min, grad, at::scalar_tensor(0., grad.options()))
+  self: where(self > min, grad, at::scalar_tensor(0., grad.options()))
   result: auto_element_wise
 
 - name: clamp_min.Tensor(Tensor self, Tensor min) -> Tensor
-  self: where(self >= min, grad, at::scalar_tensor(0., grad.options()))
-  min: where(self < min, grad, at::scalar_tensor(0., grad.options()))
-  result: where(self_p >= min_p, self_t, min_t)
+  self: at::where(self == min, grad / 2, grad).masked_fill_(self < min, 0)
```
Comment (soulitzer) [inline tools/autograd/derivatives.yaml:441]: oof this just turned 1 operator call into 4!

**Review by soulitzer — CHANGES_REQUESTED** [review]
probably regresses perf

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (HU-qingqing) [reply tools/autograd/derivatives.yaml:441]: @soulitzer @albanD   
Fair point. This mirrors the existing `maximum` / `minimum` formula so Tensor-bound clamp follows the same minimum-norm tie semantics. The extra ops are exactly why I did not change the two-Tensor-bound clamp case, where the boundary logic would be even more expensive.

If this overhead is not acceptable, the alternative seems to be a dedicated backward kernel for one-sided Tensor-bound clamp so we can keep the semantics without expanding into multiple pointwise ops. I can work on that if that sounds like the preferred direction.

**Comment by albanD** [comment]
Will be handled by https://github.com/pytorch/pytorch/pull/191142
